File: C++/14-Backtracking/Problems/2-Permutations/PermutationsPro.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
#include <set>
#include <numeric> // for iota

using namespace std;

// Function to print a vector
void printVector(const vector<int>& vec) {
    for (int num : vec) {
        cout << num << " ";
    }
    cout << endl;
}

// 1. Backtracking with Swapping (Classic Approach)
//   - Real-world use case: Generating test cases for algorithms.  You need all possible input orderings.
void permute1(vector<int>& nums, int l, int r, vector<vector<int>>& result) {
    if (l == r) {
        result.push_back(nums);
        return;
    }
    for (int i = l; i <= r; ++i) {
        swap(nums[l], nums[i]);       // Explore: Fix the current element
        permute1(nums, l + 1, r, result); // Recurse: Permute the remaining
        swap(nums[l], nums[i]);       // Backtrack: Undo the change for the next iteration
    }
}

vector<vector<int>> generatePermutations1(vector<int> nums) {
    vector<vector<int>> result;
    permute1(nums, 0, nums.size() - 1, result);
    return result;
}

// 2. Backtracking with Extra Space (Using a 'chosen' vector)
//   - Real-world use case:  Job scheduling.  Trying all possible orders to find the optimal schedule.
void permute2(vector<int>& nums, vector<int>& current, vector<bool>& chosen, vector<vector<int>>& result) {
    if (current.size() == nums.size()) {
        result.push_back(current);
        return;
    }
    for (int i = 0; i < nums.size(); ++i) {
        if (!chosen[i]) {
            chosen[i] = true;            // Mark the element as used
            current.push_back(nums[i]); // Add it to the current permutation
            permute2(nums, current, chosen, result); // Recurse
            current.pop_back();         // Backtrack: Remove the element
            chosen[i] = false;           // Backtrack: Unmark the element
        }
    }
}

vector<vector<int>> generatePermutations2(vector<int> nums) {
    vector<vector<int>> result;
    vector<int> current;
    vector<bool> chosen(nums.size(), false);
    permute2(nums, current, chosen, result);
    return result;
}

// 3. Using the STL (std::next_permutation)
//    - Real-world use case:  Generating all possible orderings for a brute-force search in a small problem space.  e.g., Trying all password combinations (within a limited set).
vector<vector<int>> generatePermutations3(vector<int> nums) {
    vector<vector<int>> result;
    sort(nums.begin(), nums.end()); // Important: Start with sorted input
    do {
        result.push_back(nums);
    } while (next_permutation(nums.begin(), nums.end())); //  Keep generating the next permutation
    return result;
}

// 4. Backtracking for Strings (Handling Duplicates)
//    - Real-world use case: Generating unique string permutations, like creating variations of a username or product code.
void permuteUnique(string& s, int l, int r, set<string>& result) { // Use a set to avoid duplicates
    if (l == r) {
        result.insert(s); // Insert into the set
        return;
    }
    for (int i = l; i <= r; ++i) {
        swap(s[l], s[i]);
        permuteUnique(s, l + 1, r, result);
        swap(s[l], s[i]); // Backtrack
    }
}

set<string> generateUniquePermutations(string s) {
    set<string> result;
    permuteUnique(s, 0, s.size() - 1, result);
    return result;
}

// 5.  Iterative Approach (Using Index Tracking) - More Complex, but avoids recursion
//     - Real-world use case:  Implementing a permutation generator in a very memory-constrained environment where recursion is too expensive.
vector<vector<int>> generatePermutations5(vector<int> nums) {
    vector<vector<int>> result;
    if (nums.empty()) return result;

    vector<int> p(nums.size(), 0); // p[i] stores the index of the element to swap with at position i
    result.push_back(nums);       // Add the original input as the first permutation
    int i = 1;
    while (i < nums.size()) {
        if (p[i] < i) {
            int j = (i % 2 == 0) ? 0 : p[i]; // Choose element to swap with
            swap(nums[i], nums[j]);
            result.push_back(nums);
            p[i]++;
            i = 1; // Reset i to 1 to start the process for the next permutation
        } else {
            p[i] = 0; // Reset p[i] if it has reached its maximum value
            i++;
        }
    }
    return result;
}

int main() {
    cout << "1. Backtracking with Swapping:\n";
    vector<int> nums1 = {1, 2, 3};
    vector<vector<int>> result1 = generatePermutations1(nums1);
    for (const auto& perm : result1) {
        printVector(perm);
    }

    cout << "\n2. Backtracking with Extra Space:\n";
    vector<int> nums2 = {1, 2, 3};
    vector<vector<int>> result2 = generatePermutations2(nums2);
    for (const auto& perm : result2) {
        printVector(perm);
    }

    cout << "\n3. Using STL (std::next_permutation):\n";
    vector<int> nums3 = {1, 2, 3};
    vector<vector<int>> result3 = generatePermutations3(nums3);
    for (const auto& perm : result3) {
        printVector(perm);
    }

    cout << "\n4. Backtracking for Strings (Handling Duplicates):\n";
    string s = "aab";
    set<string> result4 = generateUniquePermutations(s);
    for (const auto& perm : result4) {
        cout << perm << endl;
    }

    cout << "\n5. Iterative Approach:\n";
    vector<int> nums5 = {1, 2, 3};
    vector<vector<int>> result5 = generatePermutations5(nums5);
    for (const auto& perm : result5) {
        printVector(perm);
    }

    return 0;
}

File: Java/14-Backtracking/Problems/2-Permutations/PermutationsPro.java
import java.util.*;

public class Permutations {

    public static void main(String[] args) {
        int[] nums = {1, 2, 3};
        System.out.println("Input: " + Arrays.toString(nums));

        System.out.println("\nApproach 1: Backtracking with Swapping (Classic)");
        List<List<Integer>> permutations1 = generatePermutations1(nums);
        System.out.println(permutations1);

        System.out.println("\nApproach 2: Backtracking with Visited Array");
        List<List<Integer>> permutations2 = generatePermutations2(nums);
        System.out.println(permutations2);

        System.out.println("\nApproach 3: Using Collections.swap()");
        List<List<Integer>> permutations3 = generatePermutations3(nums);
        System.out.println(permutations3);

        System.out.println("\nApproach 4: Iterative Approach using Insertion");
        List<List<Integer>> permutations4 = generatePermutations4(nums);
        System.out.println(permutations4);

        System.out.println("\nApproach 5: Using a Library Function (for comparison and demonstration)");
        List<List<Integer>> permutations5 = generatePermutations5(nums);
        System.out.println(permutations5);
    }

    // Approach 1: Backtracking with Swapping (Classic)
    // This is the most common and intuitive backtracking approach.  It works by
    // fixing elements one by one and recursively generating permutations for the
    // remaining elements.  The key idea is to swap elements to explore different
    // arrangements and then backtrack to restore the original order.
    //
    // Real-world application:
    // -  Password generation:  Generating all possible password combinations.
    // -  Solving combinatorial problems:  Like finding all possible arrangements
    //    in a scheduling or resource allocation problem.
    public static List<List<Integer>> generatePermutations1(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        permute1(nums, 0, result);
        return result;
    }

    private static void permute1(int[] nums, int start, List<List<Integer>> result) {
        if (start == nums.length) {
            // Base case: When we have fixed all elements, add the current permutation to the result.
            List<Integer> permutation = new ArrayList<>();
            for (int num : nums) {
                permutation.add(num);
            }
            result.add(permutation);
            return;
        }

        for (int i = start; i < nums.length; i++) {
            // Explore all possible choices for the current position (start).
            swap(nums, start, i);            // Try placing the element at index i at the 'start' position.
            permute1(nums, start + 1, result); // Recursively generate permutations for the remaining elements.
            swap(nums, start, i);            // Backtrack: Restore the original order to explore other possibilities.
        }
    }

    private static void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }


    // Approach 2: Backtracking with Visited Array
    // This approach uses a boolean array to keep track of which elements have been
    // included in the current permutation.  It avoids swapping by explicitly
    // building the permutation step by step.
    //
    // Real-world application:
    // -  Generating unique IDs:  Ensuring that each ID is generated only once.
    // -  Solving puzzles:  Where you need to keep track of which options have
    //    already been tried (e.g., Sudoku).
    public static List<List<Integer>> generatePermutations2(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        boolean[] visited = new boolean[nums.length]; // Keep track of which numbers are used.
        List<Integer> currentPermutation = new ArrayList<>();
        permute2(nums, visited, currentPermutation, result);
        return result;
    }

    private static void permute2(int[] nums, boolean[] visited, List<Integer> currentPermutation, List<List<Integer>> result) {
        if (currentPermutation.size() == nums.length) {
            // Base case:  A complete permutation is formed.
            result.add(new ArrayList<>(currentPermutation)); // Add a *copy* of the permutation.
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (!visited[i]) {
                // If the number hasn't been used yet
                visited[i] = true;             // Mark it as used
                currentPermutation.add(nums[i]); // Add it to the current permutation
                permute2(nums, visited, currentPermutation, result); // Recurse
                currentPermutation.remove(currentPermutation.size() - 1); // Backtrack: Remove the number
                visited[i] = false;            // Mark it as unused
            }
        }
    }

    // Approach 3: Using Collections.swap()
    // This approach is similar to the first one (swapping) but uses the
    // Collections.swap() method for swapping, which can be slightly more concise.
    //
    // Real-world application:
    //   -  Shuffling algorithms:  Implementing card shuffling or randomizing lists.
    //   -  Any application where you need to rearrange elements in a list.
    public static List<List<Integer>> generatePermutations3(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        for (int num : nums) {
            list.add(num);
        }
        permute3(list, 0, result);
        return result;
    }

    private static void permute3(List<Integer> list, int start, List<List<Integer>> result) {
        if (start == list.size()) {
            result.add(new ArrayList<>(list));
            return;
        }
        for (int i = start; i < list.size(); i++) {
            Collections.swap(list, start, i);
            permute3(list, start + 1, result);
            Collections.swap(list, start, i); // Backtrack
        }
    }

    // Approach 4: Iterative Approach using Insertion
    // This approach builds permutations iteratively.  It starts with the first
    // element and then inserts each subsequent element into all possible
    // positions of the previously generated permutations.  This is a non-recursive
    // approach.
    //
    // Real-world application:
    // -  String manipulation:  Generating all possible variations of a string.
    // -  Building search indexes:  Creating different combinations of keywords.
    public static List<List<Integer>> generatePermutations4(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return result;
        }

        result.add(new ArrayList<>(Arrays.asList(nums[0]))); // Start with the first element.

        for (int i = 1; i < nums.length; i++) {
            List<List<Integer>> newResult = new ArrayList<>();
            for (List<Integer> prevPermutation : result) {
                for (int j = 0; j <= prevPermutation.size(); j++) {
                    // Insert the current number (nums[i]) at every possible position.
                    List<Integer> temp = new ArrayList<>(prevPermutation);
                    temp.add(j, nums[i]);  // Insert at index j
                    newResult.add(temp);
                }
            }
            result = newResult; // Update the result with the new permutations.
        }
        return result;
    }

    // Approach 5: Using a Library Function (for comparison and demonstration)
    // Java's Collections framework doesn't have a direct method for generating
    // permutations of an int array.  For demonstration, if you were working with
    // a List, you could adapt a library-based approach.  Guava library has
    // `Collections2.permutations()`, but here, for illustration with standard Java,
    // we show how you'd adapt the array to a List and use a conceptual approach.
    //
    // Real-world application:
    // -  Quick prototyping:  Rapidly generating permutations for testing or
    //    exploration when a library function is available.  For example, in test automation.
    public static List<List<Integer>> generatePermutations5(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        for (int num : nums) {
            list.add(num);
        }

        // In real application, if you are allowed to use external library,
        // you can use Collections2.permutations(list) from Guava.
        // Here we are implementing the basic logic.
        permute5(list, 0, result);
        return result;
    }

    private static void permute5(List<Integer> list, int start, List<List<Integer>> result) {
        if (start == list.size()) {
            result.add(new ArrayList<>(list));
            return;
        }
        for (int i = start; i < list.size(); i++) {
            Collections.swap(list, start, i);
            permute5(list, start + 1, result);
            Collections.swap(list, start, i); // Backtrack
        }
    }
}

